zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+        let mut abandoned_nodes = tree.abandoned_nodes.clone().into_inner();
+        if withdraw_all {
+            nodes[3].account = None;
+            account_to_index.remove(&account);
+            abandoned_nodes.push(3);
+        }
+
+        tree.exit(&account, amount).unwrap();
+        assert_liquidity_tree_state!(tree, nodes, account_to_index, abandoned_nodes);
+    }
+
+    #[test_case(false)]
+    #[test_case(true)]
+    fn exit_leaf_works(withdraw_all: bool) {
+        let mut tree = utility::create_test_tree();
+        // Remove lazy fees on the path to the node (and actual fees from the node).
+        tree.nodes[0].lazy_fees = Zero::zero();
+        tree.nodes[1].lazy_fees = Zero::zero();
+        tree.nodes[3].lazy_fees = Zero::zero();
+        tree.nodes[7].lazy_fees = Zero::zero();
+        tree.nodes[7].fees = Zero::zero();
+
+        let mut nodes = tree.nodes.clone().into_inner();
+        let amount = if withdraw_all { _12 } else { _1 };
+        let account = 6;
+        nodes[0].descendant_stake -= amount;
+        nodes[1].descendant_stake -= amount;
+        nodes[3].descendant_stake -= amount;
+        nodes[7].stake -= amount;
+        let mut account_to_index = tree.account_to_index.clone().into_inner();
+        let mut abandoned_nodes = tree.abandoned_nodes.clone().into_inner();
+        if withdraw_all {
+            nodes[7].account = None;
+            account_to_index.remove(&account);
+            abandoned_nodes.push(7);
+        }
+
+        tree.exit(&account, amount).unwrap();
+        assert_liquidity_tree_state!(tree, nodes, account_to_index, abandoned_nodes);
+    }
+
+    #[test_case(3, _1 + 1)]
+    #[test_case(9, _3 + 1)]
+    #[test_case(5, _3 + 1)]
+    #[test_case(7, _1 + 1)]
+    #[test_case(6, _12 + 1)]
+    #[test_case(8, _4 + 1)]
+    fn exit_fails_on_insufficient_stake(account: AccountIdOf<Runtime>, amount: BalanceOf<Runtime>) {
+        let mut tree = utility::create_test_tree();
+        // Clear unclaimed fees.
+        for node in tree.nodes.iter_mut() {
+            node.fees = Zero::zero();
+            node.lazy_fees = Zero::zero();
+        }
+        assert_err!(
+            tree.exit(&account, amount),
+            LiquidityTreeError::InsufficientStake.into_dispatch::<Runtime>(),
+        );
+    }
+
+    #[test]
+    fn exit_fails_on_unclaimed_fees_at_root() {
+        let mut tree = utility::create_test_tree();
+        // Clear unclaimed fees except for root.
+        tree.nodes[0].lazy_fees = _1;
+        tree.nodes[1].lazy_fees = Zero::zero();
+        tree.nodes[3].fees = Zero::zero();
+        tree.nodes[3].lazy_fees = Zero::zero();
+        assert_err!(tree.exit(&5, 1), LiquidityTreeError::UnclaimedFees.into_dispatch::<Runtime>());
+    }
+
+    #[test]
+    fn exit_fails_on_unclaimed_fees_on_middle_of_path() {
+        let mut tree = utility::create_test_tree();
+        // Clear unclaimed fees except for the middle node.
```

Comment:
Although the `tree` already has lazy fees of `_4`, you could write it again here in this test to make it clear to the reader, that there are outstanding unclaimed fees for node 1.

Reply:
Don't understand this. It says here that we clear unclaimed fees and set all fees to zero for these nodes. That can only mean that nodes used to have lazy fees. If the user is confused and unncessarily wishes to know exactly how many fees, they can refer to the diagram of the tree.

I mean... I'm not saying this entire test suite is beyond reproach (it clearly isn't). But virtually every test requires users to refer back to the tree and do manual calculations if they want to understand the tests. Unless I'm completely misreading this comment, I don't see why this test needs special treatment.

Reply:
So basically the idea is to add another summary of what the tree looks like before the test execution?

Reply:
I don't follow here. Making a noop assignment just to help the user not scroll down and take a look at the binary tree seems like a lot of work for virtuall zero payoff.